### PR TITLE
Distribution masking

### PR DESCRIPTION
**Pull request by H-Park**
Cleaned up setting action_mask in distribution classes, added support for other algorithms

### CODE REVIEW

File: stable_baselines/common/distributions.py [hunk 1]
```diff
@@ -293,12 +285,20 @@ def __init__(self, logits):
         :param logits: ([float]) the categorical logits input
         """
         self.logits = logits
+        self.action_mask = None
```
Comment (ChengYen-Tang) [inline stable_baselines/common/distributions.py:288]: Maybe some developers want to use Discrete.

So is this better?
class CategoricalProbabilityDistribution(ProbabilityDistribution):
    def __init__(self, logits, action_mask):

File: stable_baselines/common/distributions.py [hunk 2]
```diff
@@ -354,6 +358,10 @@ def __init__(self, nvec, flat, action_mask_vector):
         self.flat = flat
         self.categoricals = list(map(CategoricalProbabilityDistribution, tf.split(flat, nvec, axis=-1)))
 
+    def set_action_mask(self, action_mask):
+        for i, categorical in enumerate(self.categoricals):
+            categorical.set_action_mask(action_mask[:, i])
+
```
Comment (ChengYen-Tang) [inline stable_baselines/common/distributions.py:364]: I prefer to split the dataflow(action mask) in init.

Comment (H-Park) [reply stable_baselines/common/distributions.py:364]: But then that causes changing a bunch of function headers to do so, when every probability distribution class doesn't need it

File: stable_baselines/common/distributions.py (hunk 1 above)
Comment (H-Park) [reply stable_baselines/common/distributions.py:288]: Look at where the probability distribution is created ```proba_distribution_from_latent```, which is inherited from the base ```ProbabilityDistributionType```, which a few other classes use. To do this, you have to change the abstract method and add ```action_mask=None```. While it works, it isn't a great design pattern. The abstract class is being altered to allow for some child classes to have a feature, this is bad.

Comment (H-Park) [reply stable_baselines/common/distributions.py:288]: With my approach, the base class is preserved, and there is no un-needed (inherent) passing of ```action_mask=None``` to the distributions that do not support action_masking.

Comment (ChengYen-Tang) [reply stable_baselines/common/distributions.py:288]: My original idea is that since we want to add action masks, all algorithms add action masks.🤔

In my version, model.step must do action_mask_check first. If action_mask=None, I will auto generate a blank action mask.

File: stable_baselines/common/distributions.py (hunk 2 above)
Comment (ChengYen-Tang) [reply stable_baselines/common/distributions.py:364]: It seems that set_action_mask is required now, isn't it?
If this method is not call, the action_mask will be None.
Tf.add will get an error

From other perspectives, our approach seems to be the same.

File: stable_baselines/common/policies.py [hunk 3]
```diff
@@ -197,24 +200,36 @@ def step(self, obs, state=None, mask=None, action_mask=None):
...
+                action_mask = np.zeros((num_samples, len(self.ac_space.nvec), max(self.ac_space.nvec)), dtype=np.float32)
+            elif isinstance(self.ac_space, Discrete) or isinstance(self.ac_space, MultiBinary):
+                action_mask = np.zeros((num_samples, self.ac_space.n), dtype=np.float32)
+            else:
+                action_mask = np.zeros((num_samples, self.ac_space.shape[0]), dtype=np.float32)
+        else:
+            action_mask = np.array(action_mask, dtype=np.float32)
+            action_mask[action_mask <= 0] = -np.inf
+            action_mask[action_mask > 0] = 0
+        return action_mask
```
Comment (ChengYen-Tang) [inline stable_baselines/common/policies.py:231]: We all have to check the action_mask, right?

File: stable_baselines/common/distributions.py (hunk 2 above)
Comment (H-Park) [reply stable_baselines/common/distributions.py:364]: the sample method will not be called before set_action_mask gets executed, though. Look at the very next line

File: stable_baselines/common/distributions.py (hunk 1 above)
Comment (H-Park) [reply stable_baselines/common/distributions.py:288]: that is not what the stable-baselines teams want to do: https://github.com/hill-a/stable-baselines/pull/453#issuecomment-538782084

Comment (ChengYen-Tang) [reply stable_baselines/common/distributions.py:288]: Can you check the latest 2 commit for me?
I override proba_distribution_from_flat.
So, I restore the abstract class.

Comment (ChengYen-Tang) [reply stable_baselines/common/distributions.py:288]: I finally have a request, I want to contribute
Can you accept my PR and then overwrite my code?

File: stable_baselines/common/distributions.py (hunk 2 above)
Comment (ChengYen-Tang) [reply stable_baselines/common/distributions.py:364]: Me too, but I think set_action_mask will increase the time complexity.

**Comment by jimmiebtlr** [comment]
Getting this error atm when trying to run vectorized PPO2 with Discrete action space.  I will try to track down shortly.

image

**Comment by ChengYen-Tang** [comment]
@jimmiebtlr, 
I fixed this error, please re-pull my code.

File: stable_baselines/common/policies.py (hunk 3 above)
Comment (ChengYen-Tang) [reply stable_baselines/common/policies.py:231]: Also, I think this should check the shape of the action_mask.

**Comment by jimmiebtlr** [comment]
Looks like you beat me to it,  next thing I run into is
 
image

Currently looking into this

**Comment by jimmiebtlr** [comment]
Seems like it runs with correct dimensions the first run, then next time doing sampling it's not correct.   Could be something making an in-place change that isn't expected?  Will look further in a bit.

**Comment by ChengYen-Tang** [comment]
I forgot to consider that the tensorboard will record the tf graph. You can delete the tensorboard record first and he can work normally. I will think again about how to deal with this problem.

**Comment by ChengYen-Tang** [comment]
In theory, train_model does not require action masks, but tensorboard calls the mode method and the sample method in order to draw the tf graph.

**Comment by ChengYen-Tang** [comment]
Please re-pull my code.
Because my environment does not appear this. So, I am not sure if this method corrects your problem.

**Comment by jimmiebtlr** [comment]
Re-pulling fixed it.

Edit: And seems to be working beautifully.

**Comment by ChengYen-Tang** [comment]
image

File: stable_baselines/common/distributions.py (hunk 1 above)
Comment (ChengYen-Tang) [reply stable_baselines/common/distributions.py:288]: @H-Park,
Can you give me your e-mail address? Some things I want to talk to you, please.

Comment (H-Park) [reply stable_baselines/common/distributions.py:288]: hpark@innovationdx.com

**Comment by H-Park** [comment]
I'm going to merge this tonight
